include/deal.II/trilinos/nox.h

```diff
@@ -0,0 +1,1244 @@
+     *
+     * @note This function should return 0 in the case of success.
+     */
+    std::function<int(const VectorType &x)> setup_preconditioner;
+
+    /**
+     * User function that applies the Jacobian.
+     *
+     * @note The function is optional and is used in the case of certain
+     * configurations.
+     *
+     * @note This function should return 0 in the case of success.
+     */
+    std::function<int(const VectorType &x, VectorType &v)> apply_jacobian;
+
+    /**
+     * User function that applies the inverse of the Jacobian.
+     *
+     * @note The function is optional and is used in the case of certain
+     * configurations.
+     *
+     * @note This function should return 0 in the case of success.
+     */
+    std::function<
+      int(const VectorType &f, VectorType &x, const double tolerance)>
+      solve_with_jacobian;
+
+    /**
+     * User function that applies the inverse of the Jacobian and
+     * returns the numer of linear iterations the linear solver needed.
+     *
+     * @note This function should return -1 in the case of failure.
+     */
+    std::function<
+      int(const VectorType &f, VectorType &x, const double tolerance)>
+      solve_with_jacobian_and_track_n_linear_iterations;
+
+    /**
+     * User function that allows to check convergence in addition to
+     * ones checking the l2-norm and the number of iterations (see
+     * AdditionalData). It is run after each non-linear iteration.
+     *
+     * The input are the current iteration number @p i, the l2-norm
+     * @p norm_f of the residual vector, the current solution @p x,
+     * and the current residual vector @p f.
+     *
+     * @note The function is optional.
+     */
+    std::function<SolverControl::State(const unsigned int i,
+                                       const double       norm_f,
+                                       const VectorType & x,
+                                       const VectorType & f)>
+      check_iteration_status;
+
+    /**
+     * Function that allows to force to update the preconditioner in
+     * addition to AdditionalData::threshold_nonlinear_iterations. A reason
+     * for wanting to update the preconditioner is when the expected number
+     * of linear iterations exceeds.
+     *
+     * @note The function is optional. If no function is attached, this
+     * means implicitly a return value of false.
+     */
+    std::function<bool()> update_preconditioner_predicate;
+
+  private:
+    /**
+     * Additional data with basic settings.
+     */
+    AdditionalData additional_data;
+
+    /**
+     * Additional data with advanced settings. An overview of
+     * possible parameters is given at
+     * https://docs.trilinos.org/dev/packages/nox/doc/html/parameters.html.
+     */
+    const Teuchos::RCP<Teuchos::ParameterList> parameters;
+
+    /**
+     * Counter for number of (accumulated) residual evaluations.
+     */
+    mutable unsigned int n_residual_evluations;
```

Comment:
```suggestion
    mutable unsigned int n_residual_evaluations;
```

Reply:
Why do these variables actually have to be `mutable`? The class has no `const` member functions, so they should all be able to deal with regular member variables.

Reply:
I assumed that `solve()` is const. Apparently, the `solve()` functions in other classes are not const either; so I removed the mutable.